### PR TITLE
Optimize chunk reduction algorithms

### PR DESCRIPTION
**Pull request by nkoppel**
Optimizes chunk_sum, chunk_min, and chunk_max by using the [__shfl_down_sync](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#warp-shuffle-functions) intrinsic to transfer memory between threads instead of shared memory. Despite causing more contention in the output buffer, this implementation measures ~25% faster on the `sum` benchmark, and I was unable to reduce contention in a way that measurably improves performance.

### CODE REVIEW

**Comment by coreylowman** [comment]
Thanks for the PR! What GPU are you using for benchmarking? On an A10 there's not a difference on sum for me (for batchnorm/softmax is slightly slower actually).

Also any thoughts on doing block level reductions based on this article? https://developer.nvidia.com/blog/faster-parallel-reductions-kepler/

**Comment by nkoppel** [comment]
Interesting, newer gpu architectures must have faster shared memory. My gpu is a T500, an older laptop gpu.

**Comment by coreylowman** [comment]
For future reference: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/cuda/Reduce.cuh

**Comment by nkoppel** [comment]
Closing this for now, may reopen if I think of a way to apply pytorch's implementation.
